src/backend/vulkan/src/lib.rs

```diff
@@ -680,6 +735,152 @@ pub struct DeviceCreationFeatures {
+    where
+        I: Iterator<Item = &'static CStr>,
+    {
+        let mut remaining = requested_extensions.collect::<Vec<_>>();
+        let mut extensions = Vec::new();
+        let mut marked = HashSet::new();
+
+        while let Some(extension_name) = remaining.pop() {
+            if let Some(extension) = self.extensions.get(extension_name) {
+                if !extensions.contains(&extension.name) {
+                    if !extension.is_compatible_with_instance_version(instance_version) {
+                        warn!(
+                            "Extension {} (requires {:?}) is unsupported in version {:?}",
+                            extension.name.to_string_lossy(),
+                            extension.required_version,
+                            instance_version
+                        );
+                        continue;
+                    }
+
+                    if extension.is_promoted_by_instance_version(instance_version) {
+                        // This extension was promoted to core, so we shouldn't request it.
+                        debug!(
+                            "Extension {} was promoted in {:?} and is no longer explicitly required.",
+                            extension.name.to_string_lossy(),
+                            extension.promoted_version,
+                        );
+                        continue;
+                    }
+
+                    // `VK_AMD_negative_viewport_height` is obsoleted by `VK_KHR_maintenance1`, so we should try to add that instead.
+                    // Note this is the only extension we currently require that has this obsolescence deprecation state. If we gain more it may be worth refactoring `Extension` to support it.
+                    if extension.name == vk::AmdNegativeViewportHeightFn::name()
+                        && physical_device.supports_extension(vk::KhrMaintenance1Fn::name())
+                    {
+                        debug!(
+                            "Extension {} was obsoleted by {}, which is supported by the device.",
+                            extension.name.to_string_lossy(),
+                            vk::KhrMaintenance1Fn::name().to_string_lossy()
+                        );
+                        remaining.push(vk::KhrMaintenance1Fn::name());
+                        continue;
+                    }
+
+                    if physical_device.supports_extension(extension.name) {
+                        extensions.push(extension.name);
+                    } else {
+                        warn!(
+                            "Unsupported extension requested: {}",
+                            extension.name.to_string_lossy()
+                        );
+                    }
+                }
+
+                // Ensure we don't walk the dependency graph more than once.
+                if marked.contains(extension.name) {
+                    continue;
+                }
+                marked.insert(extension.name);
```

Comment:
I think `insert` should return a `bool` that you can use instead of checking for `.contains()`